Add Jdbc Source

trisberg opened the pull request:
- Using same config options as XD version: --query, --split, --update, --maxRowsPerPoll, --fixedDelay, --maxMessages

Resolves #86

trisberg committed:
Add Jdbc Source

- Using same config options as XD version: --query, --split, --update, --maxRowsPerPoll, --fixedDelay, --maxMessages

Resolves #86

ericbottard commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,122 @@
+	private JdbcSourceProperties properties;
+
+	@Autowired
+	DataSource dataSource;
```
Any reason why this (and the ones below) are package protected?

trisberg replied:
No, just forgot to add private, will fix

artembilan replied:
Actually, it doesn't matter. The `@Autowired` doesn't care about modificator.
We need `private` only for the code quality.

ericbottard commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceProperties.java:
```diff
@@ -0,0 +1,109 @@
+ * @author Thomas Risberg
+ */
+@ConfigurationProperties
+public class JdbcSourceProperties {
```
Should add additional jsr303 validation where appropriate (and tests): maxRowsPerPoll, fixedDelay

artembilan replied:
`query` is required...

artembilan replied:
Right, no reason to validate `maxRowsPerPoll`, because any `< 0` value is valid for the `JdbcPollingChannelAdapter` and we just pick up the full `SELECT` result on the output.

ericbottard commented on jdbc-source/src/test/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceIntegrationTests.java:
```diff
@@ -0,0 +1,171 @@
+	protected MessageCollector messageCollector;
+
+	@IntegrationTest(value = {"query=select id, name from test order by id"})
+	public static class DefaultBehavior extends JdbcSourceIntegrationTests {
```
I believe these inner sub classes won't be run by surefire, because they're not named `XXXTests`

trisberg replied:
You'r right, didn't think of that

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,122 @@
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties({JdbcSourceProperties.class, MaxMessagesProperties.class})
+@Import({PeriodicTriggerConfiguration.class})
```
Please, pay attention to the #146 when we are fighting to reconsider and generalize any trigger stuff.
For example that `MaxMessagesProperties` which is out of your attention here, although you have faced an issue with the `-1` for the polling adapter.
So, I recommend to merge this one after that and particular reworking.

trisberg replied:
I'll keep an eye on it

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,122 @@
+	@ConditionalOnProperty(value = "split", havingValue = "true", matchIfMissing = true)
+	public IntegrationFlow splitFlow() {
+		return IntegrationFlows.from("splitter")
+				.split("payload")
```
I guess your `payload` after the `JdbcPollingChannelAdapter` is `Collection` anyway:

```
List<?> payload = doPoll(this.sqlQueryParameterSource);
```

so, the default `.split()` (based on the `DefaultMessageSplitter`) meets your requirements.

trisberg replied:
yup

trisberg committed:
Add Jdbc Source

- addressing review comments

- adding missing private/public modifiers

- removing unused config references

- adding validation for fixedDelay property

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,122 @@
+		return IntegrationFlows.from("splitter")
+				.split("payload")
+				.channel(MessageChannels.executor(this.taskExecutor()))
+				.channel(source.output())
```
I don't think that we should worry about the `output` performance and parallelism.
The `source.output()` is pipe for the `Binder` - Redis, Kafka, Rabbit etc.

I mean that this `MessageChannels.executor()` is overhead.
And from here we even can get rid of this `IntegrationFlow` for the `split` option.

trisberg replied:
OK, I'll get rid of it

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,122 @@
+
+	@Bean
+	public PollerMetadata poller() {
+		return Pollers.fixedRate(properties.getFixedDelay(), TimeUnit.SECONDS).
```
You say `fixedRate`, but use `getFixedDelay` options. 
Inconsistency.
Although we can forget this if we will be based on the #146 , where one `@MessageSource` will save us.
As well the `PollerMetadata` from there can be used for DSL definitions.

trisberg replied:
Yes, I would agree. Inherited from XD where we do use fixed-delay poller, although I think it should have been fixed-rate in the first place. Will change to fixedDelay to stay consistent and then switch to #146 configs when available.

trisberg committed:
Add Jdbc Source

- addressing review comments

- renaming test sub-classes to end with 'Tests'

- removing unnecessary split flow and adding split() to main flow if needed

trisberg committed:
Add Jdbc Source

- addressing review comments

- switching to use FixedDelay poller

artembilan commented on jdbc-source/src/test/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceIntegrationTests.java:
```diff
@@ -0,0 +1,171 @@
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = {JdbcSourceApplication.class, EmbeddedDataSourceConfiguration.class})
+@IntegrationTest({"server.port=-1"})
```
Seems for me this doesn't work, because the particular implementation with the same `@IntegrationTest` overrides this value with that one.
That's why we use in other places `@WebIntegrationTest(randomPort = true)`.

No, this works, but with the default `8080` Tomcat port :smile:

artembilan replied:
I would like to see any feedback here.
I may not be right, but need to understand why you select this style of the config instead of `@WebIntegrationTest(randomPort = true)`.

Thank you!

ericbottard replied:
IMO, this is to avoid creating a web server (which we don't need here). The tests are already taking quite some time, no need to add burden that is not needed

artembilan replied:
Well, to avoid extra testing time we should avoid `@EnableAutoConfiguration`, which scans the classpath and populate a lot of stuff, which is, right, useless for the particular test-case.
But... Do we have other choice?

I'll check in the debug if `@IntegrationTest({"server.port=-1"})` helps to avoid starting Embedded Tomcat. 
Right now I don't believe: we have it in the CP, so `EmbeddedServletContainerAutoConfiguration` does its magic.

ericbottard replied:
Well, we still want this to test the module as it would behave "in the real world"

artembilan replied:
OK! Without `@WebIntegrationTest` we end up like:

```
DispatcherServletAutoConfiguration did not match
- @ConditionalOnClass classes found: org.springframework.web.servlet.DispatcherServlet (OnClassCondition)
- not a web application (OnWebApplicationCondition)
```

But the same result we get even without this noisy top-level `@IntegrationTest`.
In this scenario it does not have any effect: independently of that `server.port` value we don't step into the `ServerProperties` at all.

What is funny here that with the `@WebIntegrationTest({"server.port=-1"})` we end up with the:

```
2016-01-19 13:15:38.946  INFO 8796 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): -1 (http)
```

Although the Boot Docs say:

> To switch off the HTTP endpoints completely, but still create a WebApplicationContext, use server.port=-1 (this is sometimes useful for testing).

However we have the protection code in the `TomcatEmbeddedServletContainerFactory`:

```
protected void customizeConnector(Connector connector) {
        int port = (getPort() >= 0 ? getPort() : 0);
        connector.setPort(port);
```

Maybe that "endpoints disabling" doesn't relate to the Embedded Servlet Container...
Although yes: it isn't started:

```
return new TomcatEmbeddedServletContainer(tomcat, getPort() >= 0);
....
public void start() throws EmbeddedServletContainerException {
    try {
        addPreviouslyRemovedConnectors();
        Connector connector = this.tomcat.getConnector();
        if (connector != null && this.autoStart) {
            startConnector(connector);
              }
```

So, I guess we can remove this annotation from here, but for other tests when we would like to share some options for all tests, like we do in the `CassandraSinkIntegrationTests`, we don't have choice unless use `@WebIntegrationTest`. Otherwise the downstream `@IntegrationTest` overrides any properties from the upstream.

But I agree that it would be good to find some compromise between real world testing and the entire Spring Boot noise in the particular test.

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,97 @@
+			flowBuilder.split();
+		}
+		flowBuilder.channel(source.output());
+		return flowBuilder.get();
```
This config looks cool :smile: 
Although with Java 8 we would be better...
Just rhetorical comment - no need to answer.

trisberg replied:
I take cool ;)

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,97 @@
+		jdbcPollingChannelAdapter.setMaxRowsPerPoll(properties.getMaxRowsPerPoll());
+		if (properties.getUpdate() != null) {
+			jdbcPollingChannelAdapter.setUpdateSql(properties.getUpdate());
+		}
```
No reason to worry about the value. It is optional and `JdbcPollingChannelAdapter` takes care about its `null`.
I mean the `if` is redundant.
Of course, it is minor and we can live with this anyway :smile:

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,97 @@
+					@Override
+					public void accept(SourcePollingChannelAdapterSpec sourcePollingChannelAdapterSpec) {
+						sourcePollingChannelAdapterSpec
+								.autoStartup(false)
```
Looks like something has been changed in the SCDF since XD, so we don't need to worry about `auto-startup` phase and the default behavior works as is.
I mean this `.autoStartup(false)` is redundant.

trisberg replied:
Yeah, not sure about this one, just followed the ftp-source which has it - https://github.com/spring-cloud/spring-cloud-stream-modules/blob/master/ftp-source/src/main/java/org/springframework/cloud/stream/module/ftp/source/FtpSource.java#L106

artembilan replied:
I'll remove this on merge to avoid any similar discussions in the future.

artembilan commented on jdbc-source/src/main/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourceConfiguration.java:
```diff
@@ -0,0 +1,97 @@
+import org.springframework.integration.scheduling.PollerMetadata;
+import org.springframework.transaction.PlatformTransactionManager;
+
+import java.util.concurrent.TimeUnit;
```
Something wrong with imports settings?
Or does SCS team use other style?..
![imports](https://cloud.githubusercontent.com/assets/1071590/12278452/48f7b42a-b94f-11e5-9f9d-2e139d8482d5.jpg)

trisberg replied:
new IDE install, not configured properly, will fix

artembilan commented on jdbc-source/src/test/java/org/springframework/cloud/stream/module/jdbc/source/JdbcSourcePropertiesTests.java:
```diff
@@ -0,0 +1,150 @@
+		thrown.expectMessage("Field error in object 'target' on field 'query': rejected value [null]");
+		AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
+		context.register(Conf.class);
+		context.refresh();
```
`AnnotationConfigApplicationContext` is `Closeable`, so should be `close()`'d in the end, independently of test result.
At least STS barks here :smile:

trisberg committed:
Add Jdbc Source

- addressing review comments regarding import ordering

trisberg committed:
Add Jdbc Source

- addressing review comments, removing null check for update SQL

trisberg committed:
Add Jdbc Source

- addressing review comments, adding close() call for context, adding Before/After methods for context

artembilan commented:
The #146 has been deprecated.
This stuff can be reworked to the `@PollableSource` or just `@Autowired PollerMetadata defaultPoller`, if that.

trisberg commented:
A bit confused how I would replace this:

```
	@Bean
	public PollerMetadata poller() {
		return Pollers.fixedDelay(properties.getFixedDelay(), TimeUnit.SECONDS).
				maxMessagesPerPoll(properties.getMaxMessages()).
				transactional(transactionManager).get();
	}
```

artembilan commented:
See `TriggerConfiguration`, there is a bean like:
```
@Bean
public PollerMetadata defaultPoller(Trigger trigger) {
```
You should use it like:
```
@Autowired 
@Qualifier("defaultPoller")
private PollerMetadata poller;
```

Since SCS already has its onw `PollerMetadata.DEFAULT_POLLER`, we can't just use `@Autowired` for that case.
Also see the `FileSource` exactly for the sample which you are looking for.

And yes: you should remove that your `poller()` altogether :smile:

trisberg commented:
had to add @Import(TriggerConfiguration.class) and works better now. Still confused about the transactional part - not needed?

artembilan commented:
Indeed. Let's think if we really need it here.
We read data from DB, may update it and send to the `output`. May do that for each record with `split`.
That's all: our work here is done. The data is send to the `Binder` over the network.
No anything else which may impact the process!
It is just done here on the `output` channel.

trisberg commented:
SO, are you saying that we don't need transactional? What would happen if one of the update statements failed?

artembilan commented:
> What would happen if one of the update statements failed?

We have there (`JdbcPollingChannelAdapter`) only one of them according this (JDBC Source module) our settings:
````java
private Object poll() {
	List<?> payload = doPoll(this.sqlQueryParameterSource);
	if (payload.size() < 1) {
		payload = null;
	}
	if (payload != null && updateSql != null) {
		if (this.updatePerRow) {
			for (Object row : payload) {
				executeUpdateQuery(row);
			}
		}
		else {
			executeUpdateQuery(payload);
		}
	}
	return payload;
}
````
Where transaction, if that, is applied exactly around this `poll()`. Over the parent call stack, of course.
See `AbstractPollingEndpoint#createPoller()`  and `AbstractPollingEndpoint.Poller` class:
````java
while (initialized && (maxMessagesPerPoll <= 0 || count < maxMessagesPerPoll)) {
	try {
        	if (!pollingTask.call()) {
				break;
		}
	....
}
````

That `pollingTask.call()` is transaction in our case. 
In other words the tx (or any poller's advice) is applied to the single `MessageSource` invocation only.

So, having that all I really don't see reason to have the TX for single SELECT/UPDATE. That works atomically on the RDBMS. You can reach better isolation with the `SELECT .... FOR UPDATE`, I think...

trisberg committed:
Add Jdbc Source

- switch to use defaultPoller from TriggerConfiguration class

trisberg commented:
Rebased switched to use defaultPoller from TriggerConfiguration class

artembilan commented:
In addition the `sources.adoc` article must be there.
I'll show you polishing on the matter.

artembilan commented:
Here you are the polishing for your confirmation: https://github.com/artembilan/spring-cloud-stream-modules/commit/1fbb15bc17f2b1e4697ceecd0ab92ce7af49c28d

artembilan commented:
@trisberg , does your "LGTM" mean that I can merge ?

trisberg commented:
Yes, go for it :)

artembilan commented:
Merged as https://github.com/spring-cloud/spring-cloud-stream-modules/commit/48f8c2df4c9a8a05cab60c7170622ded98997823

artembilan commented:
Also see https://github.com/spring-cloud/spring-cloud-dataflow/commit/e2c10359e769c7a03626ca33449ecd2e04b1e122
